scipy/optimize/least_squares.py

```diff
@@ -0,0 +1,657 @@
+    ...                       jac_sparsity=sparsity_broyden(n))
+    >>> res_3.cost
+    4.5687161966109073e-23
+    >>> res_3.optimality
+    1.1650454296851518e-11
+    """
+    if method not in ['trf', 'dogbox', 'lm']:
+        raise ValueError("`method` must be 'trf', 'dogbox' or 'lm'.")
+
+    if tr_solver not in [None, 'exact', 'lsmr']:
+        raise ValueError("'tr_solver' must be None, 'exact' or 'lsmr'.")
+
+    if verbose not in [0, 1, 2]:
+        raise ValueError("`verbose` must be in [0, 1, 2].")
+
+    if len(bounds) != 2:
+        raise ValueError("`bounds` must contain 2 elements.")
+
+    x0 = np.atleast_1d(x0).astype(float)
+
+    if x0.ndim > 1:
+        raise ValueError("`x0` must have at most 1 dimension.")
+
+    lb, ub = prepare_bounds(bounds, x0)
+
+    if lb.shape != x0.shape or ub.shape != x0.shape:
+        raise ValueError("Inconsistent shapes between bounds and `x0`.")
+
+    if np.any(lb >= ub):
+        raise ValueError("Each lower bound mush be strictly less than each "
+                         "upper bound.")
+
+    if method == 'lm' and not np.all((lb == -np.inf) & (ub == np.inf)):
+        raise ValueError("Method 'lm' doesn't support bounds.")
+
+    if jac not in ['2-point', '3-point'] and not callable(jac):
+        raise ValueError("`jac` must be '2-point', '3-point' or callable.")
+
+    scaling = check_scaling(scaling, x0)
+
+    ftol, xtol, gtol = check_tolerance(ftol, xtol, gtol)
+
+    if not in_bounds(x0, lb, ub):
+        raise ValueError("`x0` is infeasible.")
+
+    def fun_wrapped(x):
+        return np.atleast_1d(fun(x, *args, **kwargs))
+
+    f0 = fun_wrapped(x0)
+    if f0.ndim != 1:
+        raise RuntimeError("`fun` must return at most 1-d array_like.")
+
+    n = x0.size
+    m = f0.size
+    if method == 'lm' and m < n:
+        raise ValueError("Method 'lm' doesn't work when the number of "
+                         "residuals is less than the number of variables.")
+
+    if jac in ['2-point', '3-point']:
+        if jac_sparsity is not None:
+            if method == 'lm':
+                warn("`jac_sparsity` is ignored for method='lm', dense "
+                     "differencing will be used.")
+                sparsity = None
+            else:
+                sparsity = check_jac_sparsity(jac_sparsity, n, m)
+        else:
+            sparsity = None
+
+        def jac_wrapped(x, f):
+            J = approx_derivative(
+                fun, x, rel_step=diff_step, method=jac, f0=f, bounds=bounds,
+                args=args, kwargs=kwargs, sparsity=sparsity)
+
+            if issparse(J) and tr_solver == 'exact':
+                warn("Jacobian is converted to dense for tr_solver='exact' "
+                     "even though `jac_sparsity` is provided. Consider using "
+                     "'lsmr' trust-region solver or set `jac_sparsity` to "
+                     "None.")
+                J = J.toarray()
+            elif not issparse(J):
+                J = np.atleast_2d(J)
+
```

Comment:
Because it does no harm and takes the least amount of code (most readable) :). But, as I suggested above, let's just remove these conversions and checks all together (if a user wants to convert to dense - it's trivial to do so in his code).